### PR TITLE
[GTK4 Prep] Update DocumentView to use Hdy.TabView

### PR DESCRIPTION
**Pull request by colinkiama**
DocumentView widget has been rewritten to use the `Hdy.TabView` widget. 

The majority of the work is done. There are a few more things that left to do but I haven't figured out how to do these yet (Any help would be very appreciated!):
- [x] Add "Open in a New Window" menu option to tab view in DocumentView.
- [x] Support opening drag and drop operations with tabs
  - [x] Drag tab out of window to open tab in new window
  - [x] Drag tab into window to move tab from one window to another

Part of #1157

### CODE REVIEW

File: src/Services/Document.vala [hunk 1]
```diff
@@ -113,6 +142,20 @@ namespace Scratch.Services {
             }
         }
 
+        private bool _loading = false;
```
Comment (jeremypw) [inline src/Services/Document.vala:145]: Do we need this private variable? Can we not just return `tab.page` (or false if `tab` is null).  Can `loading` be replaced entirely with `tab.working` inline?

File: src/Services/Document.vala [hunk 2]
```diff
@@ -61,10 +65,21 @@ namespace Scratch.Services {
             }
         }
 
-        public string tab_name {
+         public string tab_name {
```
Comment (jeremypw) [inline src/Services/Document.vala:68]: Do we need both `tab_name` and `title`?

File: src/Widgets/DocumentView.vala [hunk 3]
```diff
@@ -115,78 +140,61 @@ public class Scratch.Widgets.DocumentView : Granite.Widgets.DynamicNotebook {
...
+    public void duplicate_tab () {
+        if (current_document != null) {
+            new_document_from_clipboard (current_document.get_text ());
+        }
+    }
 
-        string new_text_file = _("Text file from %s:%d").printf (
-                                    timestamp.format ("%Y-%m-%d %H:%M:%S"), timestamp.get_microsecond ()
-                                );
+    public Services.Document search_for_document_in_tab (Hdy.TabPage tab) {
```
Comment (jeremypw) [inline src/Widgets/DocumentView.vala:176]: I don't understand what this does I'm afraid.  Can't we just get the document from the child of the `TabPage` parameter?

**Review by jeremypw — COMMENTED** [review]
Just a few initial comment from looking at the code.

File: src/Services/Document.vala (hunk 1 above)
Comment (colinkiama) [reply src/Services/Document.vala:145]: You're right, the private variable doesn't need to be there

File: src/Services/Document.vala (hunk 2 above)
Comment (colinkiama) [reply src/Services/Document.vala:68]: Nope. It's possible to just use `title`

File: src/Widgets/DocumentView.vala (hunk 3 above)
Comment (colinkiama) [reply src/Widgets/DocumentView.vala:176]: We can get the document from the child of the `TabPage` instead. `search_for_document_in_tab` was a supposed to be a temporary workaround for issues I had when I started working on this pull request.

File: src/Services/Document.vala (hunk 2 above)
Comment (colinkiama) [reply src/Services/Document.vala]: Only `title` is used now

File: src/Services/Document.vala (hunk 1 above)
Comment (colinkiama) [reply src/Services/Document.vala]: Private `_loading` variable is gone now

**Review by jeremypw — CHANGES_REQUESTED** [review]
Thanks for working on this. It mostly works now, just a couple of things I notice:
1) The tab background colour does not follow the "solarized light" theme (unlike `main`).  
2) The `tab_history` button is not connected up.  This button does not seem to be working in `main` either but if we are going to add it again here we might as well have it working.

**Comment by colinkiama** [comment]
I've addressed the additional feedback you had.

**Comment by colinkiama** [comment]
@jeremypw pinging because it has been a while

File: src/Widgets/DocumentView.vala [hunk 4]
```diff
@@ -482,4 +628,41 @@ public class Scratch.Widgets.DocumentView : Granite.Widgets.DynamicNotebook {
...
+        close_tab_section.append (_("Close Tabs to the Right"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_TABS_TO_RIGHT);
+        close_tab_section.append (_("Close Other Tabs"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_OTHER_TABS);
+        close_tab_section.append (_("Close Tab"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_TAB + "::");
+
+        var open_tab_section = new Menu ();
+        open_tab_section.append (_("Open in New Window"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_MOVE_TAB_TO_NEW_WINDOW);
+        open_tab_section.append (_("Duplicate"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_DUPLICATE_TAB);
+
+        menu.append_section (null, close_tab_section);
+        menu.append_section (null, open_tab_section);
```
Comment (danirabbit) [inline src/Widgets/DocumentView.vala:645]: It looks like these sections are the other way around in Files:

Screenshot from 2024-05-28 11 23 55
Screenshot from 2024-05-28 11 24 08

File: src/Widgets/DocumentView.vala [hunk 5]
```diff
@@ -482,4 +628,41 @@ public class Scratch.Widgets.DocumentView : Granite.Widgets.DynamicNotebook {
...
+        var menu = new GLib.Menu ();
+
+        var close_tab_section = new Menu ();
+        close_tab_section.append (_("Close Tabs to the Right"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_TABS_TO_RIGHT);
+        close_tab_section.append (_("Close Other Tabs"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_OTHER_TABS);
+        close_tab_section.append (_("Close Tab"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_CLOSE_TAB + "::");
+
+        var open_tab_section = new Menu ();
+        open_tab_section.append (_("Open in New Window"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_MOVE_TAB_TO_NEW_WINDOW);
+        open_tab_section.append (_("Duplicate"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_DUPLICATE_TAB);
```
Comment (danirabbit) [inline src/Widgets/DocumentView.vala:642]: This is "Duplicate Tab" in Files. Should be consistent :)
```suggestion
        open_tab_section.append (_("Duplicate Tab"), MainWindow.ACTION_PREFIX + MainWindow.ACTION_DUPLICATE_TAB);
```

**Review by danirabbit — CHANGES_REQUESTED** [review]
Looks like we're missing a couple of accelerators that are present in Files:

Screenshot from 2024-05-28 11 23 55


Otherwise this looks and works good to me!

**Review by jeremypw — CHANGES_REQUESTED** [review]
Thanks  for your continued work! The theming and closed tab button issues are now fixed and accels are showing.


One remaining behavioural issue I noticed is that the "Close Other Tabs" option is disabled when the selected tab is the first one. This should be enabled whenever there are more than one tab regardless of which is selected.

I'll use this branch for a while to check for any more regressions.

File: src/MainWindow.vala [hunk 6]
```diff
@@ -182,6 +192,7 @@ namespace Scratch {
             action_accelerators.set (ACTION_GO_TO, "<Control>i");
             action_accelerators.set (ACTION_SORT_LINES, "F5");
             action_accelerators.set (ACTION_NEW_TAB, "<Control>n");
+            action_accelerators.set (ACTION_DUPLICATE_TAB, "<Control><Alt>t" );
```
Comment (jeremypw) [inline src/MainWindow.vala:195]: This clashes with the "ACTION_TOGGLE_TERMINAL" accelerator.

Comment (colinkiama) [reply src/MainWindow.vala]: I've updated the accelerator to use `CTRL + SHIFT + K`. This matches GNOME Web and Microsoft Edge. Is this okay?

**Comment by jeremypw** [comment]
Requesting review from @danirabbit regarding what accels to use consistently for tab menu between Code, Files, Terminal and Web (if they need to be consistent).

**Comment by danirabbit** [comment]
Looks like Chrome and Edge are also using "Ctrl Shift K". Safari uses Super Shift K. So it makes sense for us to be in line with what appears to be a de facto standard here. We should probably add this shortcut to Files too

**Comment by jeremypw** [comment]
OK, I'll leave it as is here and modify Files and Terminal to match.

**Review by jeremypw — APPROVED** [review]

**Review by danirabbit — APPROVED** [review]
Let's go! :rocket:

**Comment by jeremypw** [comment]
Just found another regression - dropping a file from Files onto the tabbar (empty space) no longer opens it in a new tab.  @colinkiama   Would you like to submit a new PR fixing that? If not I'll try to fix it some time.

**Comment by colinkiama** [comment]
I won't be working on this until next week. Feel free to fix this in the meantime.

**Comment by jeremypw** [comment]
@colinkiama I could not get dropping onto the empty tabbar space to work in my Terminal branch https://github.com/elementary/terminal/pull/760 - there was no obvious way to access a widget for that area. I could only implement dropping onto the add button and tab labels.  In Code, dropping onto the add button already works.  I can implement dropping onto tab labels easily enough.

**Comment by jeremypw** [comment]
Before converting to Hdy.TabView, Code did not support dropping onto the tab labels, I guess because it is not clear what should happen? Should the contents of the dropped file replace the contents of the document (destructive), should it be appended to the document,  should the document be closed and a new one created from the dropped file, replacing the old document or should a new tab be created?

**Comment by jeremypw** [comment]
Gedit only implements drop on the add button anyway so maybe we can just leave this (unless someone can find a way implementing drop on empty space

**Pull request merged**